pip/pip-433.md

```diff
@@ -0,0 +1,172 @@
+3. Checks compatibility between two clusters when enabling namespace-level replication, which includes the following
+  a. All topics’ partitions that have been created should be the same, including `__change_events`
+  b. Auto-creation policies should be the same
+  c. Support for adjusting policies automatically when enabling namespace-level policies, and provide an API to review the changes that will be applied. By the way, this feature will be an option param of enabling namespace-level geo replication
+
+# Detailed Design
+
+### Public API
+
+#### Regarding Goal 1: "always allow the replicator to register schemas if compatible"
+
+**The original design of `pulsar-admin topics set-is-allow-auto-update-schema`**
+
+```shell
+pulsar-admin namespaces set-is-allow-auto-update-schema
+  -d, --disable         Disable schema validation enforced
+  -e, --enable          Enable schema validation enforced
+```
+
+To add a new param `--always-allow-for-replicator-if-compatible`, which means that always allow the replicator to register a new schema if compatible. The default value is `true`.
+
+#### Regarding the goal-2: "always allow the replicator to trigger a topic auto-creation"
+
+**Add new API for client LookupService**
+
+```java
+/**
+ * @param topicName it exists before the PIP.
+ * @param metadataAutoCreationEnabled, it exists before the PIP.
+ * @param expectedPartitionsAutoCreated takes effect when {@param metadataAutoCreationEnabled} is true, the broker will
+ *        create a partitioned metadata topic only if the topic does not exist and the default value of
+ *        "auto-topic-creation: num-partitions" is the same as {@param expectedPartitionsAutoCreated}. Otherwise, it
+ *        returns a topic does not exist error of the topic does not exist.
+ */
+default CompletableFuture getPartitionedTopicMetadata(TopicName topicName, boolean metadataAutoCreationEnabled, Integer expectedPartitionsAutoCreated) {
+    return getPartitionedTopicMetadata(topicName, metadataAutoCreationEnabled, false);
+}
+```
+
+**The original design of `pulsar-admin namespaces set-auto-topic-creation`**
+
+```shell
+pulsar-admin namespaces set-auto-topic-creation 
+  -d, --disable         Disable allowAutoTopicCreation on namespace
+  -e, --enable          Enable allowAutoTopicCreation on namespace
+  -n, --num-partitions=<defaultNumPartitions>
+                        The default number of partitions of a topic to be
+                          auto-created, applicable to partitioned topics only
+  -t, --type=<type>     Type of topic to be auto-created. Possible values:
+                          (partitioned, non-partitioned). Default value:
+                          non-partitioned
+                          Default: non-partitioned
+```
+
+To add a new param `--always-allow-for-replicator`, which means that always allow the replicator to trigger a topic auto-creation, including partitioned topics. The default value is `true`.
```

Comment:
Do we need this option in broker.conf?

Reply:
After this PIP, the broker-level configuration `allowAutoTopicCreation` will never affect the behavior of the replicator anymore. See also: https://github.com/apache/pulsar/pull/24485/files#diff-46571a0d56c1897159ba02ef4830473ce8de4a6d0902e3678db1eead5beaed36R143-R144

> ### Configuration
> #### Regarding the goal-1 and goal-2
> - The configuration `broker.conf -> allowAutoTopicCreation` will never affect the behavior of the replicator anymore.
> - The configuration `broker.conf -> isAllowAutoUpdateSchemaEnabled` will never affect the behavior of the replicator anymore.